crates/motsu/src/lib.rs

```diff
@@ -379,6 +387,346 @@ mod ping_pong_tests {
+
+        assert_eq!(ping_event_2.from, alice);
+        assert_eq!(ping_event_2.value, TEN + ONE);
+
+        let pong_event_1 =
+            Ponged::decode_log_data(&pong_events[0], true).unwrap();
+
+        assert_eq!(pong_event_1.from, ping.address());
+        assert_eq!(pong_event_1.value, TEN);
+
+        let pong_event_2 =
+            Ponged::decode_log_data(&pong_events[1], true).unwrap();
+
+        assert_eq!(pong_event_2.from, ping.address());
+        assert_eq!(pong_event_2.value, TEN + ONE);
+    }
+
+    #[motsu::test]
+    fn assert_revert_all_event(
+        ping: Contract<PingContract>,
+        pong: Contract<PongContract>,
+        alice: Address,
+    ) {
+        let value_revert = MAGIC_ERROR_VALUE;
+
+        // Check initial state
+        let ping_events_initial = ping.all_events();
+        let pong_events_initial = pong.all_events();
+        assert!(
+            ping_events_initial.is_empty(),
+            "Ping should have no events initially"
+        );
+        assert!(
+            pong_events_initial.is_empty(),
+            "Pong should have no events initially"
+        );
+
+        // Make a call that will revert
+        let _err = ping
+            .sender(alice)
+            .ping(pong.address(), value_revert)
+            .motsu_unwrap_err();
+
+        let ping_events_after_revert = ping.all_events();
+        let pong_events_after_revert = pong.all_events();
+
+        assert!(
+            ping_events_after_revert.is_empty(),
+            "Ping should have no events after revert"
+        );
+        assert!(
+            pong_events_after_revert.is_empty(),
+            "Pong should have no events after revert"
+        );
+
+        // make successful call
+        ping.sender(alice).ping(pong.address(), TEN).motsu_unwrap();
+
+        // Check both events
+        let ping_events = ping.all_events();
+        let pong_events = pong.all_events();
+
+        assert_eq!(ping_events.len(), 1);
+        assert_eq!(pong_events.len(), 1);
+
+        // Make another call that will revert
+        let _err = ping
+            .sender(alice)
+            .ping(pong.address(), value_revert)
+            .motsu_unwrap_err();
+
+        // Check that events are still the same
+        let ping_events_after_second_revert = ping.all_events();
+        let pong_events_after_second_revert = pong.all_events();
+
+        assert_eq!(ping_events_after_second_revert.len(), 1);
+        assert_eq!(pong_events_after_second_revert.len(), 1);
+    }
+
+    // This test checks that all events are correctly accumulated across
+    // multiple instances of the same contract type
+    #[motsu::test]
+    fn all_events_multiple_instances_same_type(
+        pong: Contract<PongContract>,
+        alice: Address,
+        bob: Address,
+    ) {
+        // Create two PingContract instances at different addresses
+        let ping1 = Contract::<PingContract>::new();
+        let ping2 = Contract::<PingContract>::new();
```

Comment:
```suggestion
    fn all_events_multiple_instances_same_type(
        ping1: Contract<PingContract>,
        ping2: Contract<PingContract>,
        pong: Contract<PongContract>,
        alice: Address,
        bob: Address,
    ) {
```
Couldn't we initiate pings like this?